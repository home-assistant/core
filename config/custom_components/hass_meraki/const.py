"""Constants for the Meraki integration."""

DOMAIN = "hass_meraki"
