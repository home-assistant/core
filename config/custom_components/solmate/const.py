"""Constants for the solmate integration."""

DOMAIN = "solmate"
