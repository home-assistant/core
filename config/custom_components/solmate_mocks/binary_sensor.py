"""Binary sensor platform for solmate_mocks integration."""

import logging

from homeassistant.components.binary_sensor import BinarySensorEntity
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(
    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
):
    """Set up the binary sensor platform."""
    _LOGGER.info("Setting up binary sensor platform")
    async_add_entities([MockFastChargeButton()])


class MockFastChargeButton(BinarySensorEntity):
    """Binary sensor for mocking fast charge button."""

    _attr_name = "Mock Fast Charge Button"
    _attr_unique_id = "mock_fast_charge_button"
    # _attr_device_class = BinarySensorDeviceClass

    def __init__(self) -> None:
        """Initialize the binary sensor."""
        self._attr_is_on = False
