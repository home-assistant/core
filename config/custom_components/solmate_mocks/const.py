"""Constants for the solmate_mocks integration."""

DOMAIN = "solmate_mocks"
