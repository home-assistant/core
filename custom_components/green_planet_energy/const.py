"""Constants for the Green Planet Energy integration."""

DOMAIN = "green_planet_energy"

# Config flow keys
CONF_API_URL = "api_url"

# Default values
DEFAULT_API_URL = "https://mein.green-planet-energy.de/p2"
DEFAULT_SCAN_INTERVAL = 60  # minutes

# API constants
API_METHOD = "getVerbrauchspreisUndWindsignal"
AGGREGATS_ZEITRAUM = ""
AGGREGATS_TYP = ""
SOURCE = "Portal"

# Sensor keys for each hour (9-18)
SENSOR_HOURS = [9, 10, 11, 12, 13, 14, 15, 16, 17, 18]
