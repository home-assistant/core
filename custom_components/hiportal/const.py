"""Constants for hiportal."""

DOMAIN = "hiportal"

CONF_USERNAME = "username"
CONF_PASSWORD = "password"
