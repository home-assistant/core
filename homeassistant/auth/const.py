"""Constants for the auth module."""

from datetime import timedelta

ACCESS_TOKEN_EXPIRATION = timedelta(minutes=30)
MFA_SESSION_EXPIRATION = timedelta(minutes=5)
REFRESH_TOKEN_EXPIRATION = timedelta(days=90).total_seconds()

GROUP_ID_ADMIN = "system-admin"
GROUP_ID_USER = "system-users"
GROUP_ID_READ_ONLY = "system-read-only"
