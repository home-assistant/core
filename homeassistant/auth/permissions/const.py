"""Permission constants."""

CAT_ENTITIES = "entities"
CAT_CONFIG_ENTRIES = "config_entries"
SUBCAT_ALL = "all"

POLICY_READ = "read"
POLICY_CONTROL = "control"
POLICY_EDIT = "edit"
