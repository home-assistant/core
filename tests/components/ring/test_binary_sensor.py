"""The tests for the Ring binary sensor platform."""

import time
from unittest.mock import Mock, patch

from freezegun.api import FrozenDateTimeFactory
import pytest
from ring_doorbell import Ring
from syrupy.assertion import SnapshotAssertion

from homeassistant.components.binary_sensor import DOMAIN as BINARY_SENSOR_DOMAIN
from homeassistant.components.ring.binary_sensor import RingEvent
from homeassistant.components.ring.const import DOMAIN
from homeassistant.components.ring.coordinator import RingEventListener
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import STATE_OFF, STATE_ON, Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers import entity_registry as er, issue_registry as ir
from homeassistant.setup import async_setup_component

from .common import MockConfigEntry, setup_automation, setup_platform
from .device_mocks import (
    FRONT_DEVICE_ID,
    FRONT_DOOR_DEVICE_ID,
    INGRESS_DEVICE_ID,
    INTERNAL_DEVICE_ID,
)

from tests.common import async_fire_time_changed, snapshot_platform


@pytest.fixture
def create_deprecated_binary_sensor_entities(
    hass: HomeAssistant,
    mock_config_entry: ConfigEntry,
    entity_registry: er.EntityRegistry,
):
    """Create the entity so it is not ignored by the deprecation check."""
    mock_config_entry.add_to_hass(hass)

    def create_entry(device_name, device_id, key):
        unique_id = f"{device_id}-{key}"

        entity_registry.async_get_or_create(
            domain=BINARY_SENSOR_DOMAIN,
            platform=DOMAIN,
            unique_id=unique_id,
            suggested_object_id=f"{device_name}_{key}",
            config_entry=mock_config_entry,
        )

    create_entry("front", FRONT_DEVICE_ID, "motion")
    create_entry("front_door", FRONT_DOOR_DEVICE_ID, "motion")
    create_entry("internal", INTERNAL_DEVICE_ID, "motion")

    create_entry("ingress", INGRESS_DEVICE_ID, "ding")
    create_entry("front_door", FRONT_DOOR_DEVICE_ID, "ding")


async def test_states(
    hass: HomeAssistant,
    mock_ring_client: Mock,
    mock_config_entry: MockConfigEntry,
    entity_registry: er.EntityRegistry,
    snapshot: SnapshotAssertion,
    create_deprecated_binary_sensor_entities,
) -> None:
    """Test states."""
    await setup_platform(hass, Platform.BINARY_SENSOR)
    await snapshot_platform(hass, entity_registry, snapshot, mock_config_entry.entry_id)


@pytest.mark.parametrize(
    ("device_id", "device_name", "alert_kind", "device_class"),
    [
        pytest.param(
            FRONT_DOOR_DEVICE_ID,
            "front_door",
            "motion",
            "motion",
            id="front_door_motion",
        ),
        pytest.param(
            FRONT_DOOR_DEVICE_ID,
            "front_door",
            "ding",
            "occupancy",
            id="front_door_ding",
        ),
        pytest.param(
            INGRESS_DEVICE_ID, "ingress", "ding", "occupancy", id="ingress_ding"
        ),
    ],
)
async def test_binary_sensor(
    hass: HomeAssistant,
    mock_config_entry: ConfigEntry,
    mock_ring_client: Ring,
    mock_ring_event_listener_class: RingEventListener,
    entity_registry: er.EntityRegistry,
    freezer: FrozenDateTimeFactory,
    device_id: int,
    device_name: str,
    alert_kind: str,
    device_class: str,
) -> None:
    """Test the Ring binary sensors."""
    # Create the entity so it is not ignored by the deprecation check
    mock_config_entry.add_to_hass(hass)

    entity_id = f"binary_sensor.{device_name}_{alert_kind}"
    unique_id = f"{device_id}-{alert_kind}"

    entity_registry.async_get_or_create(
        domain=BINARY_SENSOR_DOMAIN,
        platform=DOMAIN,
        unique_id=unique_id,
        suggested_object_id=f"{device_name}_{alert_kind}",
        config_entry=mock_config_entry,
    )
    with patch("homeassistant.components.ring.PLATFORMS", [Platform.BINARY_SENSOR]):
        assert await async_setup_component(hass, DOMAIN, {})

    on_event_cb = mock_ring_event_listener_class.return_value.add_notification_callback.call_args.args[
        0
    ]

    # Default state is set to off

    state = hass.states.get(entity_id)
    assert state is not None
    assert state.state == STATE_OFF
    assert state.attributes["device_class"] == device_class

    # A new alert sets to on
    event = RingEvent(
        1234546, device_id, "Foo", "Bar", time.time(), 180, kind=alert_kind, state=None
    )
    mock_ring_client.active_alerts.return_value = [event]
    on_event_cb(event)
    state = hass.states.get(entity_id)
    assert state is not None
    assert state.state == STATE_ON

    # Test that another event resets the expiry callback
    freezer.tick(60)
    async_fire_time_changed(hass)
    await hass.async_block_till_done()
    event = RingEvent(
        1234546, device_id, "Foo", "Bar", time.time(), 180, kind=alert_kind, state=None
    )
    mock_ring_client.active_alerts.return_value = [event]
    on_event_cb(event)
    state = hass.states.get(entity_id)
    assert state is not None
    assert state.state == STATE_ON

    freezer.tick(120)
    async_fire_time_changed(hass)
    await hass.async_block_till_done()
    state = hass.states.get(entity_id)
    assert state is not None
    assert state.state == STATE_ON

    # Test the second alert has expired
    freezer.tick(60)
    async_fire_time_changed(hass)
    await hass.async_block_till_done()
    state = hass.states.get(entity_id)
    assert state is not None
    assert state.state == STATE_OFF


async def test_binary_sensor_not_exists_with_deprecation(
    hass: HomeAssistant,
    mock_config_entry: ConfigEntry,
    mock_ring_client: Ring,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test the deprecated Ring binary sensors are deleted or raise issues."""
    mock_config_entry.add_to_hass(hass)

    entity_id = "binary_sensor.front_door_motion"

    assert not hass.states.get(entity_id)
    with patch("homeassistant.components.ring.PLATFORMS", [Platform.BINARY_SENSOR]):
        assert await async_setup_component(hass, DOMAIN, {})

    assert not entity_registry.async_get(entity_id)
    assert not er.async_entries_for_config_entry(
        entity_registry, mock_config_entry.entry_id
    )
    assert not hass.states.get(entity_id)


@pytest.mark.parametrize(
    ("entity_disabled", "entity_has_automations"),
    [
        pytest.param(False, False, id="without-automations"),
        pytest.param(False, True, id="with-automations"),
        pytest.param(True, False, id="disabled"),
    ],
)
async def test_binary_sensor_exists_with_deprecation(
    hass: HomeAssistant,
    mock_config_entry: ConfigEntry,
    mock_ring_client: Ring,
    entity_registry: er.EntityRegistry,
    issue_registry: ir.IssueRegistry,
    entity_disabled: bool,
    entity_has_automations: bool,
) -> None:
    """Test the deprecated Ring binary sensors are deleted or raise issues."""
    mock_config_entry.add_to_hass(hass)

    entity_id = "binary_sensor.front_door_motion"
    unique_id = f"{FRONT_DOOR_DEVICE_ID}-motion"
    issue_id = f"deprecated_entity_{entity_id}_automation.test_automation"

    if entity_has_automations:
        await setup_automation(hass, "test_automation", entity_id)

    entity = entity_registry.async_get_or_create(
        domain=BINARY_SENSOR_DOMAIN,
        platform=DOMAIN,
        unique_id=unique_id,
        suggested_object_id="front_door_motion",
        config_entry=mock_config_entry,
        disabled_by=er.RegistryEntryDisabler.USER if entity_disabled else None,
    )
    assert entity.entity_id == entity_id
    assert not hass.states.get(entity_id)
    with patch("homeassistant.components.ring.PLATFORMS", [Platform.BINARY_SENSOR]):
        assert await async_setup_component(hass, DOMAIN, {})

    entity = entity_registry.async_get(entity_id)
    # entity and state will be none if removed from registry
    assert (entity is None) == entity_disabled
    assert (hass.states.get(entity_id) is None) == entity_disabled

    assert (
        issue_registry.async_get_issue(DOMAIN, issue_id) is not None
    ) == entity_has_automations
