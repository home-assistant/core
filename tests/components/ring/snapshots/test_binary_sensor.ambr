# serializer version: 1
# name: test_states[binary_sensor.front_door_ding-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': None,
    'entity_id': 'binary_sensor.front_door_ding',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <BinarySensorDeviceClass.OCCUPANCY: 'occupancy'>,
    'original_icon': None,
    'original_name': 'Ding',
    'platform': 'ring',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'ding',
    'unique_id': '987654-ding',
    'unit_of_measurement': None,
  })
# ---
# name: test_states[binary_sensor.front_door_ding-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by Ring.com',
      'device_class': 'occupancy',
      'friendly_name': 'Front Door Ding',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.front_door_ding',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_states[binary_sensor.front_door_motion-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': None,
    'entity_id': 'binary_sensor.front_door_motion',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <BinarySensorDeviceClass.MOTION: 'motion'>,
    'original_icon': None,
    'original_name': 'Motion',
    'platform': 'ring',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'motion',
    'unique_id': '987654-motion',
    'unit_of_measurement': None,
  })
# ---
# name: test_states[binary_sensor.front_door_motion-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by Ring.com',
      'device_class': 'motion',
      'friendly_name': 'Front Door Motion',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.front_door_motion',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_states[binary_sensor.front_motion-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': None,
    'entity_id': 'binary_sensor.front_motion',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <BinarySensorDeviceClass.MOTION: 'motion'>,
    'original_icon': None,
    'original_name': 'Motion',
    'platform': 'ring',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'motion',
    'unique_id': '765432-motion',
    'unit_of_measurement': None,
  })
# ---
# name: test_states[binary_sensor.front_motion-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by Ring.com',
      'device_class': 'motion',
      'friendly_name': 'Front Motion',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.front_motion',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_states[binary_sensor.ingress_ding-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': None,
    'entity_id': 'binary_sensor.ingress_ding',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <BinarySensorDeviceClass.OCCUPANCY: 'occupancy'>,
    'original_icon': None,
    'original_name': 'Ding',
    'platform': 'ring',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'ding',
    'unique_id': '185036587-ding',
    'unit_of_measurement': None,
  })
# ---
# name: test_states[binary_sensor.ingress_ding-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by Ring.com',
      'device_class': 'occupancy',
      'friendly_name': 'Ingress Ding',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.ingress_ding',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_states[binary_sensor.internal_motion-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': None,
    'entity_id': 'binary_sensor.internal_motion',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': <BinarySensorDeviceClass.MOTION: 'motion'>,
    'original_icon': None,
    'original_name': 'Motion',
    'platform': 'ring',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'motion',
    'unique_id': '345678-motion',
    'unit_of_measurement': None,
  })
# ---
# name: test_states[binary_sensor.internal_motion-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'attribution': 'Data provided by Ring.com',
      'device_class': 'motion',
      'friendly_name': 'Internal Motion',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.internal_motion',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
